velox/dwio/dwrf/reader/ReaderBase.cpp

```diff
@@ -191,16 +192,27 @@ ReaderBase::ReaderBase(
 
   schema_ = std::dynamic_pointer_cast<const RowType>(
       convertType(*footer_, 0, fileColumnNamesReadAsLowerCase));
-  DWIO_ENSURE_NOT_NULL(schema_, "invalid schema");
+  VELOX_CHECK_NOT_NULL(schema_, "invalid schema");
 
   // load stripe index/footer cache
   if (cacheSize > 0) {
-    DWIO_ENSURE_EQ(format(), DwrfFormat::kDwrf);
+    VELOX_CHECK_EQ(format(), DwrfFormat::kDwrf);
     if (input_->shouldPrefetchStripes()) {
+      const uint64_t cacheOffset = fileLength_ - tailSize;
+      // NOTE: extend the cache size up to next load quantum boundary or the
+      // tail of file footer. This is to avoid the mis-aligned in-memory and
+      // ssd cache access to avoid unnecessary in-memory cache invalidation as
+      // well as ensure all the cache accesses can be covered by checksum
+      // verification if configured. We can only do checksum verification if the
+      // SSD cache read size matches the SSD cache entry size.
+      const uint64_t adjustedCacheSize = std::min(
+          tailSize,
+          bits::roundUp(cacheOffset + cacheSize, input_->loadQuantum()) -
```

Comment:
Looks like this can happen on all `CachedBufferedInput::read` calls, not just here.  Should we fix it inside `read` instead?  Something like this:

```c++
std::unique_ptr<SeekableInputStream> CachedBufferedInput::read(
    uint64_t offset,
    uint64_t length,
    LogType /*logType*/) const {
  VELOX_CHECK_LE(offset + length, fileSize_);
  const auto adjustedLength = std::min(fileSize_ - offset, bits::roundUp(offset + length, options_.loadQuantum()) - offset);
  return std::make_unique<CacheInputStream>(
      const_cast<CachedBufferedInput*>(this),
      ioStats_.get(),
      Region{offset, adjustedLength},
      input_,
      fileNum_,
      nullptr,
      TrackingId(),
      0,
      options_.loadQuantum());
}
```

Reply:
Good point. Thanks!

Reply:
We probably should fix `CacheInputStream` to differentiate these 2 lengths.  Otherwise we will need to quantize the length everything `read` is called